extension/js/common/core/crypto/key.ts

```diff
@@ -0,0 +1,238 @@
+    if (keyType === 'openpgp') {
+      return await OpenPGPKey.parse(text);
+    } else if (keyType === 'x509') {
+      return await SmimeKey.parse(text);
+    }
+    throw new Error('Unsupported key type: ' + keyType);
+  }
+
+  public static armor = (pubkey: Key): string => {
+    if (pubkey.type === 'openpgp') {
+      return OpenPGPKey.armor(pubkey);
+    } else if (pubkey.type === 'x509') {
+      return (pubkey as unknown as { raw: string }).raw;
+    } else {
+      throw new Error('Unknown pubkey type: ' + pubkey.type);
+    }
+  }
+
+  public static asPublicKey = async (pubkey: Key): Promise<Key> => {
+    // TODO: Delegate to appropriate key type
+    if (pubkey.type === 'openpgp') {
+      return await OpenPGPKey.asPublicKey(pubkey);
+    }
+    // TODO: Assuming S/MIME keys are already public: this should be fixed.
+    return pubkey;
+  }
+
+  public static expired = (key: Key): boolean => {
+    const exp = key.expiration;
+    if (!exp) {
+      return false;
+    }
+    if (exp instanceof Date) {
+      return Date.now() > exp.getTime();
+    }
+    throw new Error(`Got unexpected value for expiration: ${exp}`);
+  }
+
+  public static dateBeforeExpirationIfAlreadyExpired = (key: Key): Date | undefined => {
+    const expiration = key.expiration;
+    return expiration && KeyUtil.expired(key) ? new Date(expiration.getTime() - 1000) : undefined;
+  }
+
+  public static parseDetails = async (armored: string): Promise<{ original: string, normalized: string, keys: KeyDetails[] }> => {
+    const { normalized, keys } = await KeyUtil.normalize(armored);
+    return { original: armored, normalized, keys: await Promise.all(keys.map(PgpKey.details)) };
+  }
+
+  // todo - this should be made to tolerate smime keys
+  public static normalize = async (armored: string): Promise<{ normalized: string, keys: OpenPGP.key.Key[] }> => {
+    try {
+      let keys: OpenPGP.key.Key[] = [];
+      armored = PgpArmor.normalize(armored, 'key');
+      if (RegExp(PgpArmor.headers('publicKey', 're').begin).test(armored)) {
+        keys = (await opgp.key.readArmored(armored)).keys;
+      } else if (RegExp(PgpArmor.headers('privateKey', 're').begin).test(armored)) {
+        keys = (await opgp.key.readArmored(armored)).keys;
+      } else if (RegExp(PgpArmor.headers('encryptedMsg', 're').begin).test(armored)) {
+        keys = [new opgp.key.Key((await opgp.message.readArmored(armored)).packets)];
+      }
+      for (const k of keys) {
+        for (const u of k.users) {
+          u.otherCertifications = []; // prevent key bloat
+        }
+      }
+      return { normalized: keys.map(k => k.armor()).join('\n'), keys };
+    } catch (error) {
+      Catch.reportErr(error);
+      return { normalized: '', keys: [] };
+    }
+  }
+
+  public static getKeyType = (pubkey: string): 'openpgp' | 'x509' | 'unknown' => {
+    if (pubkey.startsWith('-----BEGIN CERTIFICATE-----')) {
+      return 'x509';
+    } else if (pubkey.startsWith('-----BEGIN PGP ')) {
```

Comment:
It may be worth to test for the full pgp header (once for public, once for private) so that pgp messages end up being `unknown`.

Reply:
Ugh, right! :facepalm: